File: 송하연/9월4주차/도넛과 막대 그래프.java
import java.util.*;
class Solution {
    
    public int[] solution(int[][] edges) {
        int[] answer = new int[4];
        
        int max = 0;
        
        //노드 최댓값
        for(int[] connect : edges){
            max = Math.max(max, Math.max(connect[0], connect[1]));
        }
        // System.out.println("max : " + max);
        
        int[] in = new int[max+1];
        int[] out = new int[max+1];
        
        //in, out 구하기
        for(int[] connect : edges){
            out[connect[0]]++;
            in[connect[1]]++; 
            
        }
        
        
        for (int i = 0; i <= max; i++) {
            //정점 
            if(in[i]==0 && out[i]>=2){
                answer[0] = i;
            }
            //막대
            if(in[i]>0 && out[i]==0){
                answer[2]++;
            }
            //8자
            if(in[i]>=2 && out[i]>=2){
                answer[3]++;
            }       
        }
        
        //도넛
        int graphNum = out[answer[0]];
        answer[1] = graphNum - answer[2] - answer[3];   

        return answer;
    }
}

File: 한승우/9월 4주차/도넛과 막대 그래프.cpp
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

const int MAX = 1000001;

//in, out degree
int degree[MAX][2];

// edges_rows는 2차원 배열 edges의 행 길이, edges_cols는 2차원 배열 edges의 열 길이입니다.
int* solution(int** edges, size_t edges_rows, size_t edges_cols) {
    // return 값은 malloc 등 동적 할당을 사용해주세요. 할당 길이는 상황에 맞게 변경해주세요.
    int* answer = (int*)malloc(4);
    int start = 0, donut = 0, bar = 0, eight = 0;
    
    for(int i = 0; i < edges_rows; i++){
        //out
        degree[edges[i][0]][1]++;
        //in
        degree[edges[i][1]][0]++;
    }
    
    for(int i = 1; i < MAX; i++){
        if(degree[i][0] == 0 && degree[i][1] >= 2) start = i;
        else if(degree[i][0] >= 2 && degree[i][1] == 2) eight++;
        else if(degree[i][0] >= 1 && degree[i][1] == 0) bar++;
    }
    
    donut = degree[start][1] - eight - bar;
    
    answer[0] = start;
    answer[1] = donut;
    answer[2] = bar;
    answer[3] = eight;
    
    return answer;
}